logger/storage/file/adapter.go

```diff
@@ -0,0 +1,130 @@
+package file
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+type adapter struct {
+	logRoot string
+	files   map[string]*os.File
+	mutex   sync.Mutex
+}
+
+// NewStorageAdapter returns a pointer to a new instance of a file-based storage.Adapter.
+func NewStorageAdapter(logRoot string) (*adapter, error) {
+	src, err := os.Stat(logRoot)
+	if err != nil {
+		return nil, fmt.Errorf("Directory %s does not exist", logRoot)
+	}
+	if !src.IsDir() {
+		return nil, fmt.Errorf("%s is not a directory", logRoot)
+	}
+	return &adapter{logRoot: logRoot, files: make(map[string]*os.File)}, nil
+}
+
+// Write adds a log message to to an app-specific log file
+func (a *adapter) Write(app string, message string) error {
+	// Check first if we might actually have to add to the map of file pointers so we can avoid
+	// waiting for / obtaining a lock unnecessarily
+	f, ok := a.files[app]
+	if !ok {
+		// Ensure only one goroutine at a time can be adding a file pointer to the map of file
+		// pointers
+		a.mutex.Lock()
+		defer a.mutex.Unlock()
```

Comment:
It might be better to push the mutex lock into getFile, which is almost always sensitive to locking.